pex/common.py

```diff
@@ -363,7 +379,45 @@ def __str__(self):
   def delete(self):
     shutil.rmtree(self.chroot)
 
-  def zip(self, filename, mode='w'):
+  def zip(self, filename, mode='w', deterministic_timestamp=False):
     with open_zip(filename, mode) as zf:
       for f in sorted(self.files()):
-        zf.write(os.path.join(self.chroot, f), arcname=f, compress_type=zipfile.ZIP_DEFLATED)
+        full_path = os.path.join(self.chroot, f)
+        if not deterministic_timestamp:
+          zf.write(
+            full_path,
+            arcname=f,
+            compress_type=zipfile.ZIP_DEFLATED
+          )
+        else:
+          # This code mostly reimplements ZipInfo.from_file(), which is not available in Python
+          # 2.7. See https://github.com/python/cpython/blob/master/Lib/zipfile.py#L495.
+          st = os.stat(full_path)
+          isdir = stat.S_ISDIR(st.st_mode)
+          # Determine arcname, i.e. the archive name.
+          arcname = f
+          arcname = os.path.normpath(os.path.splitdrive(arcname)[1])
+          while arcname[0] in (os.sep, os.altsep):
+            arcname = arcname[1:]
+          if isdir:
+            arcname += '/'
+          # Setup ZipInfo with deterministic datetime.
+          zinfo = zipfile.ZipInfo(
+            filename=arcname,
+            # NB: the constructor expects a six tuple of year-month-day-hour-minute-second.
+            date_time=DETERMINISTIC_DATETIME.timetuple()[:6]
+          )
+          zinfo.external_attr = (st.st_mode & 0xFFFF) << 16
```

Comment:
This seems like it would all more naturally house on PermPreservingZipFile suitably renamed. It already holds our custom zip logic, which is now split here in in that class, and you might take small adavantage of the co-location to re-use this perm magic which is also handled there.